### PR TITLE
Add missing tests for System.Collections

### LINKED ISSUE #6809: [ToFDisable] Port tests for: System.Collections
As part of moving out of ToF and into the open, I've been going through the "Tree of the future" tests and calculating which of those tests provide coverage that we're missing in the open tests. The goal of this is to get the open tests universally better than the ToF tests so that we can turn the ToF tests off for ProjectK runs and save us all a lot of headache diagnosing test failures from potentially outdated tests.

For System.Collections, we're missing 198 lines of coverage.

To fulfill the uncovered lines we would need to either create new tests for them or port 31 ToF tests to the open. Those tests are: 

- [ ] System.Collections\4.0.0.0\Generic\List\L_BSO2.BinarySearch_IComparer_Negative\L_BSO2.BinarySearch_IComparer_Negative.xml
- [ ] System.Collections\4.0.0.0\BitArray\BA_GST.BitArray_GetTest_Negative\BA_GST.BitArray_GetTest_Negative.xml
- [ ] System.Collections\4.0.10.0\Generic\SortedList\SortedListTests\SortedListCtorIDicComp.Constructor_IDictionary_StringComparer.RunTests.xml
- [ ] System.Collections\4.0.10.0\Generic\SortedList\SortedListTests\SortedListCtorIDicIKeyComp.Constructor_IDictionary_IKeyComparer.RunTests.xml
- [ ] System.Collections\4.0.10.0\Generic\SortedList\SortedListTests\SortedListCtorIntIKeyComp.Constructor_int_IKeyComparer.RunTests.xml
- [ ] System.Collections\4.0.0.0\Generic\SortedSet\SST.TestDuplicateAdds_Object\SST.TestDuplicateAdds_Object.xml
- [ ] System.Collections\4.0.0.0\Generic\HashSet\HS_CT.HashSetConstructor_IEn_IEq_Comparer_NotIEquatable\HS_CT.HashSetConstructor_IEn_IEq_Comparer_NotIEquatable.xml
- [ ] System.Collections\4.0.10.0\Generic\SortedList\SortedListTests\SortedListContainsValue.ContainsValue.ContainsValueMain.xml
- [ ] System.Collections\4.0.0.0\Generic\List\L_IT.ICollectionTests\L_IT.ICollectionTests.xml
- [ ] System.Collections\4.0.0.0\Generic\List\L_AT.AddNullValueWhenReference\L_AT.AddNullValueWhenReference.xml
- [ ] System.Collections\4.0.0.0\Generic\List\L_AT.NonGenericIListAddInvalidValue\L_AT.NonGenericIListAddInvalidValue.xml
- [ ] System.Collections\4.0.0.0\Generic\List\L_CTT.CopyTo_3Count\L_CTT.CopyTo_3Count.xml
- [ ] System.Collections\4.0.0.0\Generic\List\L_CTT.CopyTo_Negative3\L_CTT.CopyTo_Negative3.xml
- [ ] System.Collections\4.0.0.0\Generic\List\L_IT.IndexOf_Val\L_IT.IndexOf_Val.xml
- [ ] System.Collections\4.0.0.0\Generic\List\List_IndexTests_FullRun.XML
- [ ] System.Collections\4.0.0.0\Generic\List\L_IT.LastIndexOf_Val\L_IT.LastIndexOf_Val.xml
- [ ] System.Collections\4.0.0.0\Generic\List\List_AddTests_FullRun.XML
- [ ] System.Collections\4.0.0.0\Generic\List\L_RRT.RemoveAllTests\L_RRT.RemoveAllTests.xml
- [ ] System.Collections\4.0.0.0\Generic\List\L_ST_Test.List_SortTests.SortComparison_Tests_Negative\L_ST_Test.List_SortTests.SortComparison_Tests_Negative.xml
- [ ] System.Collections\4.0.0.0\Generic\List\L_ST_Test.List_SortTests.SortIntIntIComparer_Tests_Negative\L_ST_Test.List_SortTests.SortIntIntIComparer_Tests_Negative.xml
- [ ] System.Collections\4.0.0.0\Generic\List\L_BSO2.BinarySearch_IComparerCount_RefEven\L_BSO2.BinarySearch_IComparerCount_RefEven.xml
- [ ] System.Collections\4.0.10.0\Generic\SortedList\SortedListTests\SortedListClear.Clear.ClearMain.xml
- [ ] System.Collections\4.0.10.0\Generic\SortedList\SortedListTests\SortedListNoGenIDicEnum.Enumerator2.NOGenIDicEnumMain.xml
- [ ] System.Collections\4.0.10.0\Generic\SortedList\SortedListTests\SortedListCtorIDic.Constructor_IDictionary.RunTests.xml
- [ ] System.Collections\4.0.0.0\Generic\SortedSet\SSST.TestCopyConstructor3\SSST.TestCopyConstructor3.xml
- [ ] System.Collections\4.0.0.0\Generic\SortedSet\SortedSetSpecificTests_FullRun.XML
- [ ] System.Collections\4.0.10.0\Generic\SortedList\SortedListTests\RemoveAt.RemoveAtMain.xml
- [ ] System.Collections\4.0.0.0\Generic\SortedDictionary\SD_VC.SortedDictionary_ValueCollectionTest2\SD_VC.SortedDictionary_ValueCollectionTest2.xml
- [ ] System.Collections\4.0.0.0\Generic\SortedSet\SSST.TestUnionWithCompare\SSST.TestUnionWithCompare.xml
- [ ] System.Collections\4.0.0.0\Generic\SortedSet\SST.TestSortedSetCountAssumptions\SST.TestSortedSetCountAssumptions.xml

Coverage reports, coverage xml files, and detailed json lists of each piece of missing coverage and the corresponding ToF test that fills it are available on request.

### PR DESCRIPTION
**Pull request by ianhays**
Ports some ToF tests for System.Collections and adds some new tests to cover gaps.

resolves #6809

@stephentoub

### CODE REVIEW

File: src/System.Collections/tests/Generic/List/List.Generic.Tests.cs [hunk 1]
```diff
@@ -52,5 +52,14 @@ protected void VerifyList(List<T> list, List<T> expectedItems)
...
         #endregion
+
+        [Theory]
+        [MemberData(nameof(ValidCollectionSizes))]
+        public void CopyTo_ArgumentValidity(int count)
+        {
+            List<T> list = GenericListFactory(count);
+            Assert.Throws<ArgumentException>(() => list.CopyTo(0, new T[0], 0, count + 1));
+            Assert.Throws<ArgumentException>(() => list.CopyTo(count, new T[0], 0, 1));
+        }
```
Comment (stephentoub) [inline src/System.Collections/tests/Generic/List/List.Generic.Tests.cs]: There aren't any other [Fact]s or [Theory]s on this class... seems like it's purely a bunch of helpers.  Is this the right place for this?

File: src/System.Collections/tests/Generic/SortedList/SortedList.Generic.Tests.cs [hunk 2]
```diff
@@ -431,6 +431,20 @@ public void SortedList_Generic_RemoveAt_EveryValue(int count)
             }
         }
 
+        [Theory]
+        [MemberData("ValidCollectionSizes")]
```
Comment (stephentoub) [inline src/System.Collections/tests/Generic/SortedList/SortedList.Generic.Tests.cs]: I'm not sure how this file escaped our replacing these with usage of nameof, but we should follow-up to fix that.  For now, it makes sense for you to remain consistent with the rest of the file and leave this as-is.

**Comment by stephentoub** [comment]
LGTM

File: src/System.Collections/tests/Generic/List/List.Generic.Tests.cs (hunk 1 above)
Comment (ianhays) [inline src/System.Collections/tests/Generic/List/List.Generic.Tests.cs]: Yeah. Most methods for this class have their own file but I don't want to have to make a whole file for one test. If anything all of those files should be merged into this one.

File: src/System.Collections/tests/Generic/SortedList/SortedList.Generic.Tests.cs (hunk 2 above)
Comment (ianhays) [inline src/System.Collections/tests/Generic/SortedList/SortedList.Generic.Tests.cs]: I'll follow up with the change now.

**Pull request merged**